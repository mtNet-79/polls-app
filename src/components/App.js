import { handleInitialData } from "../actions/shared";
import PollPage from "./PollPage";
import { useState, useEffect, Fragment } from "react";
import Dashboard from "./Dashboard";
import { connect } from "react-redux";
import Nav from './Nav';
import LogIn from "./LogIn";
import Leaderboard from "./Leaderboard";
import AddUser from "./AddUser";
import LoadingBar from "react-redux-loading-bar";
import { Routes, Route } from "react-router-dom";
import CreatePoll from "./CreatePoll";
import { useLocation, useNavigate, useParams } from "react-router-dom";

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

const App = (props) => {
  console.log("app props: ", props);
  const [image, setImage] = useState("");
  useEffect(() => {
   
    const generateRandomAvatar = () => {
      fetch("https://avatar-endpoint.herokuapp.com/api/").then((res) => {
        res.blob().then((blobRes) => {
          const data = blobRes;
          const urlCreator = window.URL || window.webkitURL;
          const urlObj = urlCreator.createObjectURL(data);
          setImage(urlObj);
          console.log("trigged in generateavatar");
          props.dispatch(handleInitialData());
        });
      });
    }

    const alreadyLoggedIn = () => {
      console.log("triggerd at already")
      props.dispatch(handleInitialData());
    }
    
    props.authedUser === null 
      ? generateRandomAvatar() 
      : alreadyLoggedIn();
  }, []);



  return (
    <Fragment>
      <LoadingBar />       
        {props.loading ? null : (   
            <div className="container">              
              {props.location.pathname === "/add-user" 
                ? <AddUser image={image} />
                : props.authedUser
                  ? (
                    <Fragment>                   
                      <Nav />
                      <Routes>
                        <Route path="/" exact element={<Dashboard />} />                        
                        <Route path="/login" element={<LogIn image={image} />} />
                        <Route path="/new" element={<CreatePoll />} />
                        <Route path="/leaderboard" element={<Leaderboard />} />
                        <Route path="/questions/:pid" element={<PollPage image={image}/>} />
                      </Routes>
                    </Fragment>   
                  )           
                  : <LogIn image={image} />                  
              }            
            </div>
        )}
    </Fragment>
  );
};

const mapStateToProps = ({  authedUser, loading, users }, props) => {  
  // const keys = Object.keys(users);
  const { location } = props.router;
  return {
    loading: loading,
    authedUser,
    avatarURL: users[authedUser]?.avatarURL,
    location,
  }
    
    
};
  

export default withRouter(connect(mapStateToProps)(App));
