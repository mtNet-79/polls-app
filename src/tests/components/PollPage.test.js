import { screen, waitFor, within } from "@testing-library/react";
// import PollContainer from "../../components/PollContainer";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import App from "../../Components/App";
import { renderWithProviders } from "./test-utils";
import userEvent from "@testing-library/user-event";

// import { formatDate } from "../../utils/helpers";

//unit tests
test("renders / take snapshot", async () => {
  const user = userEvent.setup();
  const {container} =  renderWithProviders(<App />, {
    route: "/",
    authedUser: "mthornton",
  });

  await waitFor(
    () => expect(screen.getByText(/new polls/i)).toBeInTheDocument(),
    {
      timeout: 2000,
    }
  );
  
  //all unanswered polls are present

  const dashboardList = screen.getAllByRole('list')[2]
  const pollBoxes = within(dashboardList).getAllByRole("listitem");

  expect(pollBoxes.length).toBeGreaterThan(0);

  const pollBox = pollBoxes[0];
  expect(
    within(pollBox).getByRole("button", { name: /show poll/i })
  ).toBeInTheDocument();

  await user.click(within(pollBox).getByRole("button", { name: /show poll/i }));


  expect(screen.getByRole('heading', { name: /poll by/i })).toBeInTheDocument()
  expect(screen.getByRole('img', { name: /avatar of /i })).toBeInTheDocument()
  const buttons = screen.getAllByRole('button', { name: /click/i })
  expect(buttons.length).toBe(2);
  expect(container).toMatchSnapshot();

  //TODO:click canswer

  await user.click(buttons[0])
});
