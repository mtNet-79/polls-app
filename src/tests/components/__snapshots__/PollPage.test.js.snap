// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders / take snapshot 1`] = `
<div>
  <div
    style="direction: ltr;"
  >
    <div
      class=""
      style="width: 100%; transition: width 200ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 1;"
    />
    <div
      style="display: table; clear: both;"
    />
  </div>
  <div
    class="container"
  >
    <div
      class="nav"
    >
      <ul>
        <li>
          <a
            class=""
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class=""
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li>
          <a
            class=""
            href="/new"
          >
            Post
          </a>
        </li>
      </ul>
      <ul>
        <li>
          <img
            alt="icon"
            class="avatar-small"
            src="https://avatar-endpoint.herokuapp.com/api/"
          />
          <span>
            mthornton
          </span>
        </li>
        <li>
          <a
            href="/login"
          >
            Log out
          </a>
        </li>
      </ul>
    </div>
    <h3>
      Poll by 
      mtsamis
    </h3>
    <img
      alt="Avatar of mtsamis"
      class="avatar center"
    />
    <div
      class="flex-row poll"
    >
      <div
        class="choice-container"
      >
        <div
          class="center questionBox"
        >
          deploy to production once every two weeks
        </div>
        <button
          class="btn-answer"
          id="optionOne"
        >
          Click
        </button>
      </div>
      <div
        class="choice-container"
      >
        <div
          class="center questionBox"
        >
          deploy to production once every month
        </div>
        <button
          class="btn-answer"
          id="optionOne"
        >
          Click
        </button>
      </div>
    </div>
  </div>
</div>
`;
